pallets/rewards/src/mechanism/gap.rs

```diff
@@ -0,0 +1,377 @@
+pub struct Group<T: Config> {
+	total_stake: T::Balance,
+	pending_total_stake: T::Balance,
+	rpt: T::Rate,
+	distribution_id: T::DistributionId,
+}
+
+impl<T: Config> Default for Group<T> {
+	fn default() -> Self {
+		Self {
+			total_stake: T::Balance::zero(),
+			pending_total_stake: T::Balance::zero(),
+			rpt: T::Rate::zero(),
+			distribution_id: T::DistributionId::default(),
+		}
+	}
+}
+
+#[derive(Encode, Decode, TypeInfo, MaxEncodedLen, RuntimeDebug)]
+#[cfg_attr(test, derive(PartialEq, Clone))]
+pub struct Account<T: Config> {
+	stake: T::Balance,
+	reward_tally: T::IBalance,
+	pending_stake: T::Balance,
+	distribution_id: T::DistributionId,
+	last_currency_movement: u16,
+}
+
+impl<T: Config> Default for Account<T> {
+	fn default() -> Self {
+		Self {
+			stake: T::Balance::zero(),
+			reward_tally: T::IBalance::zero(),
+			pending_stake: T::Balance::zero(),
+			distribution_id: T::DistributionId::default(),
+			last_currency_movement: 0,
+		}
+	}
+}
+
+impl<T: Config> Account<T> {
+	fn was_distribution(&self, group: &Group<T>, currency: &Currency<T>) -> bool {
+		if self.last_currency_movement as usize == currency.rpt_changes.len() {
+			self.distribution_id != group.distribution_id
+		} else {
+			self.distribution_id != currency.prev_distribution_id
+				|| group.distribution_id != currency.next_distribution_id
+		}
+	}
+
+	fn reward_tally_updated(
+		&self,
+		group: &Group<T>,
+		currency: &Currency<T>,
+	) -> Result<T::IBalance, DispatchError> {
+		let reward_tally = if self.was_distribution(group, currency) {
+			let correct_rpt = RptHistory::<T>::get(self.distribution_id).ok_or(
+				DispatchError::Other("'DistributionId' not found in 'RptHistory'"),
+			)?;
+
+			self.reward_tally.ensure_add(
+				correct_rpt
+					.ensure_mul_int(self.pending_stake)?
+					.ensure_into()?,
+			)?
+		} else {
+			self.reward_tally
+		};
+
+		let tally_rpt_changes = self.get_tally_from_rpt_changes(&currency.rpt_changes)?;
+		Ok(reward_tally.ensure_add(tally_rpt_changes)?)
+	}
+
+	fn stake_updated(
+		&self,
+		group: &Group<T>,
+		currency: &Currency<T>,
+	) -> Result<T::Balance, ArithmeticError> {
+		if self.was_distribution(group, currency) {
+			self.stake.ensure_add(self.pending_stake)
+		} else {
+			Ok(self.stake)
+		}
+	}
+
+	fn update(&mut self, group: &Group<T>, currency: &Currency<T>) -> Result<(), DispatchError> {
+		if self.was_distribution(group, currency) {
+			let stake = self.stake_updated(group, currency)?;
+			let reward_tally = self.reward_tally_updated(group, currency)?;
+
+			self.stake = stake;
+			self.reward_tally = reward_tally;
+			self.pending_stake = T::Balance::zero();
+		}
+		self.apply_rpt_changes(&currency.rpt_changes)?;
+		self.distribution_id = group.distribution_id;
+
+		Ok(())
+	}
+
+	fn get_tally_from_rpt_changes(
+		&self,
+		rpt_changes: &[T::Rate],
+	) -> Result<T::IBalance, ArithmeticError> {
+		let rpt_to_apply = &rpt_changes[self.last_currency_movement as usize..]
```

Comment:
In the worst case the call will iterate the value stored in MaxMovements, right? Do you know what value we have set this to?

Reply:
Correct. It's set at configuration. Currently, in the development environment, It's set to 50.